sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -522,6 +522,8 @@ object ColumnPruning extends Rule[LogicalPlan] {
    * so remove it.
    */
   private def removeProjectBeforeFilter(plan: LogicalPlan): LogicalPlan = plan transform {
+    case p1 @ Project(_, _ @ Filter(condition, _ @ Project(_, _: LeafNode)))
+      if !condition.deterministic => p1
```

Comment:
I don't get it from your explanation. If I understand it correctly, when there is a `Project` which selects subset of output from the `LeafNode`, if we remove it by the below pattern, we will retrieve all fields. Is it your purpose?